tensorboard/data/dispatching_provider.py

```diff
@@ -0,0 +1,211 @@
+
+        Raises:
+          ValueError: If any of the provider keys contains a colon,
+            which would make it impossible to match.
+        """
+        self._providers = dict(providers)
+        invalid_names = sorted(k for k in self._providers if _SEPARATOR in k)
+        if invalid_names:
+            raise ValueError("Invalid provider key(s): %r" % invalid_names)
+        self._unprefixed_provider = unprefixed_provider
+
+    def _parse_eid(self, experiment_id):
+        """Parse an experiment ID into prefix, sub-ID, and sub-provider.
+
+        The returned prefix may be `None` if this instance has an
+        unprefixed data provider registered. If the experiment ID is
+        invalid, this method may raise an `errors.NotFoundError`.
+        """
+        parts = experiment_id.split(_SEPARATOR, 1)
+        if len(parts) == 1:
+            if self._unprefixed_provider is None:
+                raise errors.NotFoundError(
+                    "No unprefixed data provider specified"
+                )
+            return (None, experiment_id, self._unprefixed_provider)
+        (prefix, sub_eid) = parts
+        sub_provider = self._providers.get(prefix)
+        if sub_provider is None:
+            raise errors.NotFoundError("Unknown data provider key: %r" % prefix)
+        return (prefix, sub_eid, sub_provider)
+
+    def _simple_delegate(get_method):
+        """Dispatch on experiment ID, forwarding args and result unchanged."""
+
+        def wrapper(self, *args, experiment_id, **kwargs):
+            (_, sub_eid, sub_provider) = self._parse_eid(experiment_id)
+            method = get_method(sub_provider)
+            return method(*args, experiment_id=sub_eid, **kwargs)
+
+        return wrapper
+
+    data_location = _simple_delegate(lambda p: p.data_location)
+    experiment_metadata = _simple_delegate(lambda p: p.experiment_metadata)
+    list_plugins = _simple_delegate(lambda p: p.list_plugins)
+    list_runs = _simple_delegate(lambda p: p.list_runs)
+    list_scalars = _simple_delegate(lambda p: p.list_scalars)
+    read_scalars = _simple_delegate(lambda p: p.read_scalars)
+    list_tensors = _simple_delegate(lambda p: p.list_tensors)
+    read_tensors = _simple_delegate(lambda p: p.read_tensors)
+    list_blob_sequences = _simple_delegate(lambda p: p.list_blob_sequences)
+
+    def read_blob_sequences(self, *args, experiment_id, **kwargs):
+        (prefix, sub_eid, sub_provider) = self._parse_eid(experiment_id)
+        result = sub_provider.read_blob_sequences(
+            *args, experiment_id=sub_eid, **kwargs
+        )
+        for tag_to_data in result.values():
+            for (tag, old_data) in tag_to_data.items():
+                new_data = [
+                    provider.BlobSequenceDatum(
+                        step=d.step,
+                        wall_time=d.wall_time,
+                        values=_convert_blob_references(prefix, d.values),
```

Comment:
I’m open to inlining this method if reviewers think that that would be
clearer. It’d drop a top-level method and docstring and shorten the code
overall, at the cost of a nested loop comprehension here.